### PR TITLE
8327111: Replace remaining usage of create_bool_from_template_assertion_predicate() which requires additional OpaqueLoop*Nodes transformation strategies

### PR DESCRIPTION
**Pull request by chhagedorn**
https://github.com/openjdk/jdk/pull/18293 started to replace `create_bool_from_template_assertion_predicate()` usages to fix an endless DFS traversal problem. This patch is a follow-up to replace the last usage of `create_bool_from_template_assertion_predicate()` in `clone_assertion_predicate_and_initialize()` to completely fix the problem.

Depending on where `clone_assertion_predicate_and_initialize()` is called from, we need to clone the Template Assertion Predicate Expression differently:
- Create a new Template Assertion Predicate for a main loop: Clone everything except for the `OpaqueLoopInitNode` which needs to be replaced with a new `OpaqueLoopInitNode` (done with `clone_and_replace_init()`).
- Create an Initialized Assertion Predicate for all other cases: Clone everything except for the `OpaqueLoop*Nodes` which are replaced with an actual init and stride value (done with `clone_and_replace_init_and_stride()`).

Note that it's incorrect to _not_ create new Template Assertion Predicates in case we split the loop further (for example after peeling a loop). This will be eventually fixed with [JDK-8288981](https://bugs.openjdk.org/browse/JDK-8288981).

I've extended the test added with https://github.com/openjdk/jdk/pull/18293 to also cover the now fixed cases. 

Thanks,
Christian

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8327111](https://bugs.openjdk.org/browse/JDK-8327111): Replace remaining usage of create_bool_from_template_assertion_predicate() which requires additional OpaqueLoop*Nodes transformation strategies (**Sub-task** - P4)

### Reviewers
 * [Emanuel Peter](https://openjdk.org/census#epeter) (@eme64 - **Reviewer**)
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18628/head:pull/18628` \
`$ git checkout pull/18628`

Update a local copy of the PR: \
`$ git checkout pull/18628` \
`$ git pull https://git.openjdk.org/jdk.git pull/18628/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18628`

View PR using the GUI difftool: \
`$ git pr show -t 18628`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18628.diff">https://git.openjdk.org/jdk/pull/18628.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18628#issuecomment-2039089228)

### CODE REVIEW

File: src/hotspot/share/opto/loopTransform.cpp [hunk 1]
```diff
@@ -1488,96 +1488,23 @@ void PhaseIdealLoop::count_opaque_loop_nodes(Node* n, uint& init, uint& stride)
...
 // Clone an Assertion Predicate for the main loop. new_init and new_stride are set as new inputs. Since the predicates
 // cannot fail at runtime, Halt nodes are inserted instead of uncommon traps.
 Node* PhaseIdealLoop::clone_assertion_predicate_and_initialize(Node* iff, Node* new_init, Node* new_stride, Node* predicate,
                                                                Node* uncommon_proj, Node* control, IdealLoopTree* outer_loop,
                                                                Node* input_proj) {
-  Node* result = create_bool_from_template_assertion_predicate(iff, new_init, new_stride, control);
+  TemplateAssertionPredicateExpression template_assertion_predicate_expression(iff->in(1)->as_Opaque4());
+  Opaque4Node* new_opaque4_node;
+  if (new_stride == nullptr) {
+    new_opaque4_node = template_assertion_predicate_expression.clone_and_replace_init(new_init, control, this);
```
Comment (eme64) [inline src/hotspot/share/opto/loopTransform.cpp:1499]: Optional: Maybe you could add a small comment about why the `new_stride` is a `nullptr` in this case?

**Review by eme64 — APPROVED** [review]
Looks good, a nice continuation of your series of fixes and refactorings!

File: src/hotspot/share/opto/loopTransform.cpp (hunk 1 above)
Comment (chhagedorn) [reply src/hotspot/share/opto/loopTransform.cpp:1502]: Thanks for your review! Good idea, added a comment.

Comment (eme64) [reply src/hotspot/share/opto/loopTransform.cpp:1502]: Ah very nice, even an assert!

**Review by eme64 — APPROVED** [review]

**Comment by chhagedorn** [comment]
Thanks @eme64 for your review!

**Review by vnkozlov — APPROVED** [review]
Looks good.

**Comment by chhagedorn** [comment]
Thanks @vnkozlov for your review!

I will now merge everything into https://github.com/openjdk/jdk/pull/16877 and update the PR accordingly with the new code of JDK-8271109/8271110/8271111.

/integrate

**Pull request closed without merging**